pkg/controller/master-controller-manager/kubeone/controller.go

```diff
@@ -122,19 +200,121 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 		return reconcile.Result{}, err
 	}
 
-	err := r.reconcile(ctx, externalCluster)
+	if externalCluster.DeletionTimestamp != nil {
+		log.Debug("Deleting KubeOne Namespace")
+		ns := &corev1.Namespace{}
+		name := types.NamespacedName{Name: kubernetesprovider.GetKubeOneNameSpaceName(externalCluster.Name)}
+		if err := r.Get(ctx, name, ns); err != nil {
+			return reconcile.Result{}, err
+		}
+		if err := r.Delete(ctx, ns); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if err := kuberneteshelper.TryRemoveFinalizer(ctx, r, externalCluster, apiv1.ExternalClusterKubeOneNamespaceCleanupFinalizer); err != nil {
+			log.Errorf("Could not delete kubeone namespace, %w", err)
+			return reconcile.Result{}, err
+		}
+	}
+	return r.reconcile(ctx, externalCluster)
+}
+
+func (r *reconciler) reconcile(ctx context.Context, externalCluster *kubermaticv1.ExternalCluster) (reconcile.Result, error) {
+	kubeconfigSecret := &corev1.Secret{}
+	if err := r.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: kubernetesprovider.GetKubeOneNameSpaceName(externalCluster.Name), Name: resources.KubeOneKubeconfigSecretName}, kubeconfigSecret); err != nil {
+		if kerrors.IsNotFound(err) {
+			if err := r.importCluster(ctx, externalCluster); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		return reconcile.Result{}, err
+	}
+	externalClusterProvider, err := kubernetesprovider.NewExternalClusterProvider(r.ImpersonationClient, r.Client)
 	if err != nil {
-		log.Errorw("Reconciling failed", zap.Error(err))
+		return reconcile.Result{}, err
+	}
+	version, err := externalClusterProvider.GetVersion(ctx, externalCluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	currentVersion := version.String()
+
+	manifestSecret := &corev1.Secret{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: externalCluster.Spec.CloudSpec.KubeOne.ManifestReference.Namespace, Name: externalCluster.Spec.CloudSpec.KubeOne.ManifestReference.Name}, manifestSecret); err != nil {
+		return reconcile.Result{}, fmt.Errorf("can not retrieve kubeone manifest secret: %w", err)
+	}
+	currentManifest := manifestSecret.Data[resources.KubeOneManifest]
+	cluster := &kubeonev1beta2.KubeOneCluster{}
+	if err := yaml.UnmarshalStrict(currentManifest, cluster); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to decode manifest secret data: %w", err)
+	}
+
+	wantedVersion := cluster.Versions.Kubernetes
+	// check if a upgrade pod already exists
+	pod := &corev1.Pod{}
+	err = r.Get(ctx, ctrlruntimeclient.ObjectKey{
+		Name:      KubeOneUpgradePod,
+		Namespace: kubernetesprovider.GetKubeOneNameSpaceName(externalCluster.Name),
+	},
+		pod,
+	)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			if externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status != kubermaticv1.StatusError && currentVersion != wantedVersion {
+				upgradeMsg := fmt.Sprintf("Upgrading kubeone cluster from %s=>%s", currentVersion, wantedVersion)
+				r.log.Debugw(upgradeMsg)
```

Comment:
Please use structured logging correctly, e.g. `r.log.Debugw("Upgrading cluster", "from" currentVersion, "to", wantedVersion)`